drivers/network/netio/netio.c

```diff
@@ -0,0 +1,1454 @@
+
+    Some minor functions (not used by WinDRBD) are missing
+    (like WskControlClient, WskSocketConnect, ...)
+    Raw sockets are not supported for now.
+    We should have regression tests for that...
+*/
+
+#include <ntifs.h>
+#include <netiodef.h>
+#include <ws2def.h>
+#include <wsk.h>
+#include <ndis.h>
+#define DBG 1
+#include <netio_debug.h>
+
+#include <tdi.h>
+#include <tcpioctl.h>
+#include <tdikrnl.h>
+#include <tdiinfo.h>
+#include <tdi_proto.h>
+#include <tdiconn.h>
+
+ULONG DebugTraceLevel = MIN_TRACE;
+
+#define TAG_NETIO 'OIEN'
+
+#ifndef TAG_AFD_TDI_CONNECTION_INFORMATION
+#define TAG_AFD_TDI_CONNECTION_INFORMATION 'cTfA'
+#endif
+
+/* AFD Share Flags */
+#define AFD_SHARE_UNIQUE    0x0L
+#define AFD_SHARE_REUSE     0x1L
+#define AFD_SHARE_WILDCARD  0x2L
+#define AFD_SHARE_EXCLUSIVE 0x3L
+
+/* Function trace */
+// #define FUNCTION_TRACE DbgPrint("Function %s ...\n", __func__)
+#define FUNCTION_TRACE do { } while (0)
+
+#if 0
+
+/* Very simple malloc debugger .. I would like to keep this here
+ * for now in case we find another BSOD.
+ */
+
+#define ExAllocatePoolWithTag(pool, size, tag) \
+({ \
+	void *ret; \
+	if (tag == TAG_NETIO) { \
+		ret = ExAllocatePoolWithTag(pool, size+4, tag); \
+		if (ret != NULL) \
+			*(int*)(((char*)ret)+size) = 0xabcdef; \
+	} else { \
+		ret = ExAllocatePoolWithTag(pool, size, tag); \
+	} \
+        NETIO_DbgPrint(MIN_TRACE, ("ExAllocatePoolWithTag returning %p size is %d tag is %x\n", ret, size, tag)) \
+	ret; \
+})
+
+#define ExFreePoolWithTag(p, tag) \
+	NETIO_DbgPrint(MIN_TRACE, ("ExFreePoolWithTag %p %x\n", p, tag)) \
+	ExFreePoolWithTag(p, tag)
+
+#endif
+
+typedef struct _WSK_SOCKET_INTERNAL
+{
+    WSK_SOCKET s;
+
+    ADDRESS_FAMILY family;      /* AF_INET or AF_INET6 */
+    USHORT type;                /* SOCK_DGRAM, SOCK_STREAM, ... */
+    ULONG proto;                /* IPPROTO_UDP, IPPROTO_TCP */
+    ULONG flags;                /* WSK_FLAG_LISTEN_SOCKET, ... */
+    PVOID user_context;         /* parameter for callbacks, opaque */
+    UNICODE_STRING TdiName;     /* \\Devices\\Tcp, \\Devices\\Udp */
+
+    struct sockaddr LocalAddress;
+    PFILE_OBJECT LocalAddressFile;
+    HANDLE LocalAddressHandle;
+
+    struct sockaddr RemoteAddress;      /* Remeber latest remote connection */
+    PFILE_OBJECT RemoteAddressFile;
+    HANDLE RemoteAddressHandle;
+
+    /* Those exist for connection oriented (TCP/IP) sockets only */
+    PFILE_OBJECT ConnectionFile;        /* Returned by TdiOpenConnectionEndpointFile() */
+    HANDLE ConnectionHandle;            /* Returned by TdiOpenConnectionEndpointFile() */
+    BOOLEAN ConnectionFileAssociated;
+
+    /* Incoming connection callback function: */
+    const WSK_CLIENT_LISTEN_DISPATCH *ListenDispatch;
+    UINT CallbackMask;
+
+    UINT Flags;                          /* SO_REUSEADDR, ... see ws2def.h */
+    UINT RefCount;                       /* See SocketGet/SocketPut TODO: this should be atomic */
```

Comment:
Please fix (see the suggestions that follow.)
```suggestion
    LONG RefCount;                       /* See SocketGet/SocketPut */
```
changed to "LONG" for using the InterlockedIncrement/Decrement functions.